akka-projection-slick/src/test/scala/akka/projection/slick/SlickProjectionSpec.scala

```diff
@@ -0,0 +1,215 @@
+    super.beforeAll()
+    repository.createTable()
+  }
+
+  "A Slick projection" must {
+
+    "persist events and offset in same transactional" in {
+
+      implicit val dispatcher = actorSystem.dispatcher
+
+      val projectionId = UUID.randomUUID().toString
+      val entityId = UUID.randomUUID().toString
+
+      withClue("check - offset is empty") {
+        val offsetOpt = offsetStore.readOffset[Long](projectionId).futureValue
+        offsetOpt shouldBe empty
+      }
+
+      val slickProjection =
+        SlickProjection.transactional(
+          projectionId = projectionId,
+          sourceProvider = sourceProvider(entityId),
+          offsetExtractor = offsetExtractor,
+          databaseConfig = dbConfig) { envelope => repository.concatToText(envelope.id, envelope.message) }
+
+      runProjection(slickProjection) {
+        eventually {
+          withClue("check - all values were concatenated") {
+            val concatStr = dbConfig.db.run(repository.findById(entityId)).futureValue.value
+            concatStr.text shouldBe "abc|def|ghi|jkl|mno|pqr"
+          }
+        }
+      }
+      withClue("check - all offsets were seen") {
+        val offset = offsetStore.readOffset[Long](projectionId).futureValue.value
+        offset shouldBe 6L
+      }
+    }
+
+    "restart from previous offset " in {
+
+      implicit val dispatcher = actorSystem.dispatcher
+
+      val projectionId = UUID.randomUUID().toString
+      val entityId = UUID.randomUUID().toString
+
+      withClue("check - offset is empty") {
+        val offsetOpt = offsetStore.readOffset[Long](projectionId).futureValue
+        offsetOpt shouldBe empty
+      }
+
+      val slickProjectionFailing =
+        SlickProjection.transactional(
+          projectionId = projectionId,
+          sourceProvider = sourceProvider(entityId),
+          offsetExtractor = offsetExtractor,
+          databaseConfig = dbConfig) { envelope =>
+          if (envelope.offset == 4L) DBIOAction.failed(new RuntimeException("fail on third element"))
+          else repository.concatToText(envelope.id, envelope.message)
+        }
+
+      runProjection(slickProjectionFailing) {
+        eventually {
+          withClue("check: projection is interrupted on fourth element") {
+            val concatStr = dbConfig.db.run(repository.findById(entityId)).futureValue.value
+            concatStr.text shouldBe "abc|def|ghi"
+          }
+        }
+      }
+      withClue("check: last seen offset is 3L") {
+        val offset = offsetStore.readOffset[Long](projectionId).futureValue.value
+        offset shouldBe 3L
+      }
+
+      // re-run projection without failing function
+      val slickProjection =
+        SlickProjection.transactional(
+          projectionId = projectionId,
+          sourceProvider = sourceProvider(entityId),
+          offsetExtractor = offsetExtractor,
+          databaseConfig = dbConfig) { envelope => repository.concatToText(envelope.id, envelope.message) }
```

Comment:
This is one of the reasons I would like to review this API. This reads to me as a bunch of code thrown on the editor. First a bunch of configurations/boilerplate and then the user function.

Maybe it will be better to have a class: EventHandler or EventProcessor. User code tend to be more complex, having a function will either lead people to throw here lots of code or to create their own EventHandler class and lift to a function here. 

Any suggestions are welcome.

Reply:
Another advantage of an `EventHandler` class is that we can add some error handling strategies. The user defines their logic and error handling in one place, following some pre-defined API. That will make it easier for us to manipulate it behind.

Reply:
Doesn't look too bad for this simple event handler, but worth exploring.

I think a class also encourages that the handler can have state. That's ofc possible with a function also, but less obvious that it s safe since most (should) think a function is stateless. We could always provide a convenience implementation of such class that just wraps a function.

Reply:
Create an issue to explore it later. 
https://github.com/akka/akka-projection/issues/43